### PR TITLE
🗻 Add config reader, rework components, Engine main

### PR DESCRIPTION
**Pull request by RenardFute**
This pull request involves significant refactoring and cleanup within the `src/common` directory, focusing on removing deprecated files and adding new subdirectories to better organize the codebase. Key changes include the removal of several classes and the addition of new directories in the CMake configuration, as well as updates to the `IComponent` and `IObject` interfaces.

#### Code Cleanup and Removal:
* Removed `AComponent`, `AField`, `AMeta`, and their associated header files, simplifying the codebase by eliminating unused or deprecated classes. (`src/common/AComponent.cpp` [[1]](diffhunk://#diff-86381311cee8d358b69d05b331f9a4ea0a02bf916eaf02718e0a6d0b97edf3afL1-L40) `src/common/AField.cpp` [[2]](diffhunk://#diff-0831262199525f0584640bded62d2a83c8b60e917edc71ea452bd094496dbe3eL1-L30) `src/common/AMeta.cpp` [[3]](diffhunk://#diff-89372bea01a72da30878c83a90b3aec2dec54b2b34656bf500cfaad97f953960L1-L43) `src/common/AMeta.hpp` [[4]](diffhunk://#diff-beb8b2c6dfc166850f3e51e30fcfe26b49439846e5da935f994b4c3fbcc9673dL1-L138)

#### CMake Configuration:
* Updated `CMakeLists.txt` to add new subdirectories (`factories`, `fields`, `components`, `event`, `managers`) and removed references to deleted files. (`src/common/CMakeLists.txt` [src/common/CMakeLists.txtR6-R28](diffhunk://#diff-623b734f767b84dff28cb49ebd68b44d56a1a0eabcc436fed6cd0ed74eb7b19fR6-R28))

#### Interface Updates:
* Enhanced `IComponent` interface by adding new methods (`onCreation`, `onDeletion`, `updateValue`, `getValue`, `serialize`) and modifying existing ones to improve flexibility and functionality. (`src/common/IComponent.hpp` [[1]](diffhunk://#diff-501b2cab76de290d4794d7ed087e653e88b402eb087827fab7f1578b75271e85L36-R61) [[2]](diffhunk://#diff-501b2cab76de290d4794d7ed087e653e88b402eb087827fab7f1578b75271e85R159-R161) [[3]](diffhunk://#diff-501b2cab76de290d4794d7ed087e653e88b402eb087827fab7f1578b75271e85R236-R255) [[4]](diffhunk://#diff-501b2cab76de290d4794d7ed087e653e88b402eb087827fab7f1578b75271e85L275-R319) [[5]](diffhunk://#diff-501b2cab76de290d4794d7ed087e653e88b402eb087827fab7f1578b75271e85R332-R350)
* Refined `IObject` interface by removing copy constructors and assignment operators, enforcing better object management practices. (`src/common/IObject.hpp` [[1]](diffhunk://#diff-4a36ae616a4b46000f5a79ea2107c8a7372a6fe250df8b6709b6fb0f9c6f56d3L34-L62) [[2]](diffhunk://#diff-4a36ae616a4b46000f5a79ea2107c8a7372a6fe250df8b6709b6fb0f9c6f56d3L193-L234)

#### New Classes:
* Introduced `VirtualObject` and `VirtualScene` classes to handle object and scene management, respectively, providing a more modular and testable architecture. (`src/common/VirtualObject.cpp` [[1]](diffhunk://#diff-18242410eb3325675b3af4da4e969789c45d0d1a78b41db9c520dc97343aab87R1-R104) `src/common/VirtualObject.hpp` [[2]](diffhunk://#diff-b1437b638a7c45761004e3ca051f8889f23c2d0fc68eac3e9e4170eaf502c3a5R1-R65) `src/common/VirtualScene.cpp` [[3]](diffhunk://#diff-57c9a68214ae3708d337155022c6c656c22ceff486d706bfd35eaef50f480d12R1-R44)

#### Utility Functions:
* Added a utility header `Utils.hpp` with a template function to get the type name of an object, aiding in debugging and type identification. (`src/common/Utils.hpp` [src/common/Utils.hppR1-R18](diffhunk://#diff-82167a2975ab3d08197f51dc701c6a070a19270ca155a0d5ae5e255612512f3fR1-R18))

These changes collectively aim to streamline the codebase, improve maintainability, and enhance the functionality of core interfaces and components.

### CODE REVIEW

File: src/common/factories/ComponentFactory.cpp
Comment (Marius-P1) [inline src/common/factories/ComponentFactory.cpp:1]: You can delete this if it's not used

File: src/common/factories/ComponentFactory.hpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    [[nodiscard]]
+    static IComponent *create(const std::string &typeName,
+                              IObject *owner,
+                              const json::JsonObject *data) {
+        if (registry().find(typeName) != registry().end()) {
+            return registry()[typeName](owner, data);
+        }
+        return nullptr;
```
Comment (Marius-P1) [inline src/common/factories/ComponentFactory.hpp:45]: You can put this into the cpp

File: src/common/json/JsonObject.cpp [hunk 2]
```diff
@@ -89,3 +89,8 @@ void JsonObject::setValue(const std::string &key, IJsonObject *value) {
     this->_objects[key] = value;
 }
 
+bool JsonObject::contains(const std::string &key) const {
+    return this->_objects.find(key) != this->_objects.end();
+}
+
+
```
Comment (Marius-P1) [inline src/common/json/JsonObject.cpp:96]:
```suggestion
```

File: src/graphics/components/AnimatedSprite.cpp [hunk 3]
```diff
@@ -7,89 +7,159 @@
...
-{
+#include "common/components/Transform.hpp"
+#include "common/fields/FileField.hpp"
+#include "common/json/JsonArray.hpp"
+#include "common/json/JsonNumber.hpp"
+#include "common/json/JsonString.hpp"
+
+namespace json {
+    class JsonString;
+}
```
Comment (Marius-P1) [inline src/graphics/components/AnimatedSprite.cpp:18]: Is the forward mandatory (you have already include the header)

File: src/graphics/components/SpriteSheet.cpp [hunk 4]
```diff
@@ -5,74 +5,129 @@
 ** SpriteSheet.cpp
 **/
 
+#include <SFML/Graphics/Rect.hpp>
+
```
Comment (Marius-P1) [inline src/graphics/components/SpriteSheet.cpp:9]:
```suggestion
```

File: src/graphics/components/SpriteSheet.cpp [hunk 5]
```diff
@@ -5,74 +5,129 @@
...
 #include "SpriteSheet.hpp"
```
Comment (Marius-P1) [inline src/graphics/components/SpriteSheet.cpp:11]:
```suggestion
```

File: src/physics/components/RigidBody.cpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
+/*
+** EPITECH PROJECT, 2024
+** StellarForge
+** File description:
+** No file there, just an epitech header example.
+*/
+
+#include "RigidBody.hpp"
+
```
Comment (Marius-P1) [inline src/physics/components/RigidBody.cpp:9]:
```suggestion
```

File: src/physics/components/RigidBody.cpp [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+RigidBody::Meta::getFieldGroups() const {
+    return {&_fieldGroup};
+}
+
+json::IJsonObject *RigidBody::serializeData() {
+    return new json::JsonNull();
+}
+
+void RigidBody::deserialize(const json::IJsonObject *data) {
+}
```
Comment (Marius-P1) [inline src/physics/components/RigidBody.cpp:78]: It's normal that the method are empty ?

**Review by Marius-P1 — APPROVED** [review]
Except the lack of documentation, it seem to be working :see_no_evil:

File: src/physics/components/RigidBody.cpp (hunk 7 above)
Comment (RenardFute) [reply src/physics/components/RigidBody.cpp:78]: Yes seen with @landryarki

**Pull request merged**